# 🛒 Indian Retail Store — Sales Analysis
**Dataset:** 100,000 transactions | 2019–2023  
**Tools:** Python, Pandas  
**Goal:** Identify sales trends, profit drivers & customer behavior

In [1]:
# libraries needed
import pandas as pd
import numpy as np

# Data Cleaning 

In [3]:
# import dataset as data
data = pd.read_csv("C:/Users/deepi/Downloads/store_sales_data.csv")

In [4]:
# view data
data

,Customer ID,Customer Name,Last Name,Date of Birth,Sales,Year,Outlet Type,City Type,Category of Goods,Region,...,Ship Date,Ship Mode,State,Postal Code,Product ID,Sub-Category,Product Name,Quantity,Discount,Profit
0,CUST000001,Curtis,Krause,1973-11-29,3344.69,2023,Large,Tier 1,Fast Food,East,...,2020-04-18,Same Day,West Bengal,700001,PROD000001,Burgers,Burgers - 718,3,0.09,448.24
1,CUST000002,Jeremy,Foster,1987-08-30,37094.41,2022,Medium,Tier 2,Fast Food,East,...,2021-01-11,Standard Class,Punjab,160001,PROD000002,Pizzas,Pizzas - 519,5,0.39,3134.34
2,CUST000003,Ariel,Mccann,1980-12-15,35501.08,2023,Large,Village,Electric Appliances,South,...,2020-08-16,Second Class,Gujarat,380001,PROD000003,Fans,Fans - 395,6,0.07,8324.21
3,CUST000004,Colleen,Maynard,1989-05-23,8969.33,2021,Medium,Village,Dairy Products,East,...,2021-10-16,First Class,West Bengal,700001,PROD000004,Milk,Milk - 798,4,0.18,1962.66
4,CUST000005,Gina,Holden,1957-12-31,35530.44,2021,Small,Village,Sessional Fruits & Vegetables,West,...,2019-01-25,Standard Class,Delhi,110001,PROD000005,Tomatoes,Tomatoes - 442,4,0.27,2843.93
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
99995,CUST099996,Thomas,Adams,2018-02-02,254.35,2020,Large,Tier 1,Furniture,East,...,2019-02-05,First Class,Gujarat,380001,PROD099996,Beds,Beds - 169,2,0.12,61.84
99996,CUST099997,Clifford,Dillon,1955-08-11,846.66,2019,Medium,Tier 2,Sessional Fruits & Vegetables,North,...,2020-04-09,Same Day,Rajasthan,302001,PROD099997,Tomatoes,Tomatoes - 291,1,0.43,118.60
99997,CUST099998,Kelly,Deleon,1991-02-27,12165.34,2019,Large,Tier 2,Furniture,East,...,2021-07-02,Standard Class,Maharashtra,400001,PROD099998,Chairs,Chairs - 381,4,0.13,2100.20
99998,CUST099999,Katie,Moses,1990-10-21,41242.99,2019,Small,Tier 1,Fast Food,North,...,2019-08-01,Second Class,Karnataka,560001,PROD099999,Sandwiches,Sandwiches - 961,4,0.41,4674.57


In [5]:
# size of the Dataset
data.shape

(100000, 25)

In [6]:
# datatypes of each column
data.dtypes

Customer ID           object
Customer Name         object
Last Name             object
Date of Birth         object
Sales                float64
Year                   int64
Outlet Type           object
City Type             object
Category of Goods     object
Region                object
Country               object
Segment               object
Sales Date            object
Order ID              object
Order Date            object
Ship Date             object
Ship Mode             object
State                 object
Postal Code            int64
Product ID            object
Sub-Category          object
Product Name          object
Quantity               int64
Discount             float64
Profit               float64
dtype: object

In [7]:
# null count
data.isnull().sum()

Customer ID          0
Customer Name        0
Last Name            0
Date of Birth        0
Sales                0
Year                 0
Outlet Type          0
City Type            0
Category of Goods    0
Region               0
Country              0
Segment              0
Sales Date           0
Order ID             0
Order Date           0
Ship Date            0
Ship Mode            0
State                0
Postal Code          0
Product ID           0
Sub-Category         0
Product Name         0
Quantity             0
Discount             0
Profit               0
dtype: int64

In [8]:
# duplicated values count
data.duplicated().sum()

0

In [9]:
# changing datatype for date columns
date_cols = ['Order Date','Date of Birth','Sales Date','Ship Date']
for cols in date_cols:
    data[cols] = pd.to_datetime(data[cols])

In [10]:
# re-checking datatypes after converting
data.dtypes

Customer ID                  object
Customer Name                object
Last Name                    object
Date of Birth        datetime64[ns]
Sales                       float64
Year                          int64
Outlet Type                  object
City Type                    object
Category of Goods            object
Region                       object
Country                      object
Segment                      object
Sales Date           datetime64[ns]
Order ID                     object
Order Date           datetime64[ns]
Ship Date            datetime64[ns]
Ship Mode                    object
State                        object
Postal Code                   int64
Product ID                   object
Sub-Category                 object
Product Name                 object
Quantity                      int64
Discount                    float64
Profit                      float64
dtype: object

In [11]:
# Created columns like Profit margin % and Ship days
data['Profit Margin %'] = (data['Profit'] / data['Sales'] * 100).round(2)
data['Ship Days'] = (data['Ship Date'] - data['Order Date']).dt.days

In [12]:
# sum of sales
data['Sales'].sum()

2508441014.18

In [13]:
# Categories
data['Category of Goods'].unique()

array(['Fast Food', 'Electric Appliances', 'Dairy Products',
       'Sessional Fruits & Vegetables', 'Furniture', 'Household Items'],
      dtype=object)

In [14]:
# View dataset after adding columns
data

,Customer ID,Customer Name,Last Name,Date of Birth,Sales,Year,Outlet Type,City Type,Category of Goods,Region,...,State,Postal Code,Product ID,Sub-Category,Product Name,Quantity,Discount,Profit,Profit Margin %,Ship Days
0,CUST000001,Curtis,Krause,1973-11-29,3344.69,2023,Large,Tier 1,Fast Food,East,...,West Bengal,700001,PROD000001,Burgers,Burgers - 718,3,0.09,448.24,13.40,5
1,CUST000002,Jeremy,Foster,1987-08-30,37094.41,2022,Medium,Tier 2,Fast Food,East,...,Punjab,160001,PROD000002,Pizzas,Pizzas - 519,5,0.39,3134.34,8.45,3
2,CUST000003,Ariel,Mccann,1980-12-15,35501.08,2023,Large,Village,Electric Appliances,South,...,Gujarat,380001,PROD000003,Fans,Fans - 395,6,0.07,8324.21,23.45,5
3,CUST000004,Colleen,Maynard,1989-05-23,8969.33,2021,Medium,Village,Dairy Products,East,...,West Bengal,700001,PROD000004,Milk,Milk - 798,4,0.18,1962.66,21.88,6
4,CUST000005,Gina,Holden,1957-12-31,35530.44,2021,Small,Village,Sessional Fruits & Vegetables,West,...,Delhi,110001,PROD000005,Tomatoes,Tomatoes - 442,4,0.27,2843.93,8.00,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
99995,CUST099996,Thomas,Adams,2018-02-02,254.35,2020,Large,Tier 1,Furniture,East,...,Gujarat,380001,PROD099996,Beds,Beds - 169,2,0.12,61.84,24.31,5
99996,CUST099997,Clifford,Dillon,1955-08-11,846.66,2019,Medium,Tier 2,Sessional Fruits & Vegetables,North,...,Rajasthan,302001,PROD099997,Tomatoes,Tomatoes - 291,1,0.43,118.60,14.01,7
99997,CUST099998,Kelly,Deleon,1991-02-27,12165.34,2019,Large,Tier 2,Furniture,East,...,Maharashtra,400001,PROD099998,Chairs,Chairs - 381,4,0.13,2100.20,17.26,5
99998,CUST099999,Katie,Moses,1990-10-21,41242.99,2019,Small,Tier 1,Fast Food,North,...,Karnataka,560001,PROD099999,Sandwiches,Sandwiches - 961,4,0.41,4674.57,11.33,5


# Exploratory Data Analysis

In [16]:
# Statistical view of the important columns
data[['Sales','Profit','Quantity','Discount','Ship Days']].describe()

,Sales,Profit,Quantity,Discount,Ship Days
count,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000
mean,25084.410142,3755.305114,5.493760,0.251320,3.991060
std,14403.187686,2639.852035,2.876102,0.144324,2.006554
min,100.570000,6.520000,1.000000,0.000000,1.000000
25%,12618.030000,1651.105000,3.000000,0.130000,2.000000
50%,25134.695000,3317.450000,5.000000,0.250000,4.000000
75%,37575.850000,5363.845000,8.000000,0.380000,6.000000
max,49999.890000,14563.670000,10.000000,0.500000,7.000000


In [17]:
# grouping sales by  categories
data.groupby('Category of Goods')['Sales'].sum().round(2)
# profit by each category
data.groupby('Category of Goods')['Profit'].sum().round(2)

Category of Goods
Dairy Products                   62592431.61
Electric Appliances              62702992.93
Fast Food                        62787837.10
Furniture                        61809104.10
Household Items                  62871293.87
Sessional Fruits & Vegetables    62766851.82
Name: Profit, dtype: float64

In [18]:
data.groupby('Category of Goods').agg({
    'Sales':'sum',
    'Profit':'sum'
}).sort_values('Sales',ascending = False) /1e6

,Sales,Profit
Category of Goods,,
Sessional Fruits & Vegetables,419.942695,62.766852
Household Items,419.788895,62.871294
Electric Appliances,418.829608,62.702993
Dairy Products,418.011671,62.592432
Fast Food,417.850836,62.787837
Furniture,414.017308,61.809104


In [19]:
# region-wise sales, profit and orders count
result = data.groupby('Region').agg({
    'Sales':'sum',
    'Profit':'sum',
    'Order ID':'count'
})
result[['Sales','Profit']] = result[['Sales','Profit']] / 1e6
result.sort_values('Sales',ascending = False).round(2)

,Sales,Profit,Order ID
Region,,,
East,635.32,95.15,25117
South,629.06,94.22,25031
North,624.45,93.10,25043
West,619.61,93.06,24809


In [20]:
# Year 
data.groupby('Year').agg({
    'Sales':'sum',
    'Profit':'sum'
}).sort_values('Sales',ascending = False) / 1e6

,Sales,Profit
Year,,
2021,504.754579,75.358055
2020,504.242259,75.384910
2019,503.069477,75.533395
2022,498.858828,74.799411
2023,497.515871,74.454740


In [21]:
data.groupby('Outlet Type').agg({
    'Profit Margin %':'mean',
    'Profit':'sum'
}).sort_values('Profit Margin %',ascending = False).round(2)

,Profit Margin %,Profit
Outlet Type,,
Medium,14.98,1.261441e+08
Small,14.98,1.239613e+08
Large,14.93,1.254250e+08


### 📌 EDA Findings
- **Dataset** : 100,000 rows, 25 columns, 0 nulls
- **Categories** : 6 — all roughly equal in sales
- **Regions** : East leads, all close
- **Years** : Peak 2021, declining since
- **Outlets** : Size doesn't affect margin

# Sales Trend Analysis

In [24]:
# extract time columns
data['Order Year'] = data['Order Date'].dt.year
data['Order Month'] = data['Order Date'].dt.month
data['Order Quarter'] = data['Order Date'].dt.quarter

In [25]:
# monthly sales trends
data.groupby(['Order Year','Order Month'])['Sales'].sum()

Order Year  Order Month
2019        1              42344783.81
            2              37865145.00
            3              43913954.31
            4              40725028.79
            5              42081183.23
            6              41175051.76
            7              43041403.30
            8              42947691.21
            9              42853719.63
            10             42585146.28
            11             40655408.34
            12             41989205.76
2020        1              43335493.32
            2              39734929.06
            3              41219136.88
            4              41501540.88
            5              41949577.55
            6              40424208.29
            7              42678848.46
            8              41171210.53
            9              41028621.71
            10             44325510.82
            11             41953290.59
            12             41434554.57
2021        1              44244055.65
 

In [26]:
monthly = data.groupby(['Order Year','Order Month'])['Sales'].sum() / 1e6
print("Best Month: ", monthly.idxmax())
print("Worst Month: ",monthly.idxmin())

Best Month:  (2023, 5)
Worst Month:  (2019, 2)


In [27]:
data.groupby('Order Quarter')['Sales'].sum() / 1e6

Order Quarter
1    619.712576
2    626.832590
3    632.390375
4    629.505473
Name: Sales, dtype: float64

# Observation

### 📌 Sales Trend Findings

- Sales peaked in **2021** at ₹504.7M and have declined since
- **May (Month 5)** is the best performing month every year
- **February (Month 2)** is the weakest month every year — consistent across all 5 years
- February dips range between ₹37.8M – ₹39.8M every year
- Monthly sales range between ₹37M and ₹44M
- **Q3 (Jul–Sep)** is the strongest quarter at ₹632M
- All quarters are balanced — no extreme seasonality

In [30]:
data.head()

,Customer ID,Customer Name,Last Name,Date of Birth,Sales,Year,Outlet Type,City Type,Category of Goods,Region,...,Sub-Category,Product Name,Quantity,Discount,Profit,Profit Margin %,Ship Days,Order Year,Order Month,Order Quarter
0,CUST000001,Curtis,Krause,1973-11-29,3344.69,2023,Large,Tier 1,Fast Food,East,...,Burgers,Burgers - 718,3,0.09,448.24,13.40,5,2020,4,2
1,CUST000002,Jeremy,Foster,1987-08-30,37094.41,2022,Medium,Tier 2,Fast Food,East,...,Pizzas,Pizzas - 519,5,0.39,3134.34,8.45,3,2021,1,1
2,CUST000003,Ariel,Mccann,1980-12-15,35501.08,2023,Large,Village,Electric Appliances,South,...,Fans,Fans - 395,6,0.07,8324.21,23.45,5,2020,8,3
3,CUST000004,Colleen,Maynard,1989-05-23,8969.33,2021,Medium,Village,Dairy Products,East,...,Milk,Milk - 798,4,0.18,1962.66,21.88,6,2021,10,4
4,CUST000005,Gina,Holden,1957-12-31,35530.44,2021,Small,Village,Sessional Fruits & Vegetables,West,...,Tomatoes,Tomatoes - 442,4,0.27,2843.93,8.00,3,2019,1,1


In [31]:
data.columns.tolist()

['Customer ID',
 'Customer Name',
 'Last Name',
 'Date of Birth',
 'Sales',
 'Year',
 'Outlet Type',
 'City Type',
 'Category of Goods',
 'Region',
 'Country',
 'Segment',
 'Sales Date',
 'Order ID',
 'Order Date',
 'Ship Date',
 'Ship Mode',
 'State',
 'Postal Code',
 'Product ID',
 'Sub-Category',
 'Product Name',
 'Quantity',
 'Discount',
 'Profit',
 'Profit Margin %',
 'Ship Days',
 'Order Year',
 'Order Month',
 'Order Quarter']

In [32]:
result = data.groupby('Sub-Category').agg({
    'Sales'          : 'sum',
    'Profit'         : 'sum',
    'Profit Margin %': 'mean'
}).sort_values('Sales', ascending=False).head(10).round(2)

result[['Sales','Profit']] = result[['Sales','Profit']] / 1e6
print(result)

                       Sales     Profit  Profit Margin %
Sub-Category                                            
Mops              109.568527  16.395505            15.03
Carrots           107.517944  16.136519            14.98
Milk              105.820115  15.782967            14.89
Microwaves        105.807952  16.083505            15.20
Pizzas            105.703349  15.817570            14.83
Sofas             105.501807  15.787824            14.96
Burgers           105.356521  15.829860            15.02
Cheese            105.335995  15.770492            14.99
Refrigerators     105.112393  15.650139            14.96
Washing Machines  104.901980  15.564815            14.83


### Category & Regional Findings
- **Mops** is the top sub-category at ₹109.5M
- Top 10 sub-categories are tightly clustered (₹104M–₹109M)
- **East region** leads overall sales
- **Electric Appliances** has the highest profit margin at 14.99%

In [34]:
result = data.groupby('Category of Goods').agg({
    'Profit Margin %': 'mean',
    'Sales'          : 'sum',
    'Profit'         : 'sum'
}).sort_values('Profit Margin %', ascending=False).round(2)

result[['Sales','Profit']] = result[['Sales','Profit']] / 1e6
print(result)

                               Profit Margin %       Sales     Profit
Category of Goods                                                    
Electric Appliances                      14.99  418.829608  62.702993
Dairy Products                           14.98  418.011671  62.592432
Household Items                          14.98  419.788895  62.871294
Fast Food                                14.97  417.850836  62.787837
Sessional Fruits & Vegetables            14.97  419.942695  62.766852
Furniture                                14.90  414.017308  61.809104


- **Key insight** → Furniture has the lowest profit margin at 14.90% despite being a high-value category. This could be due to higher discounts or operational costs.
- **All margins are extremely close** — between 14.90% and 14.99%. Only 0.09% difference across all categories.

### Category & Regional Findings

- **Electric Appliances** has the highest profit margin at 14.99%
- **Furniture** has the lowest margin at 14.90% — needs investigation
- All categories are tightly clustered within 0.09% margin range
- **Mops** is the top sub-category by sales at ₹109.5M
- Pricing is consistent across all categories — no major outliers

# Discount Impact & Customer Segmentation

Do higher discounts lead to lower profit margins?

In [38]:
# discount band
data['Discount Band'] = pd.cut(
    data['Discount'],
    bins=[-0.01, 0.0, 0.1, 0.2, 0.3, 0.4, 0.5],
    labels=['No Discount','1-10%','11-20%','21-30%','31-40%','41-50%']
)

In [39]:
result = data.groupby('Discount Band', observed=True).agg({
    'Profit Margin %': 'mean',
    'Sales'          : 'sum',
    'Order ID'       : 'count'
}).round(2)
result[['Sales']] = result[['Sales']] / 1e6
print(result)

               Profit Margin %       Sales  Order ID
Discount Band                                       
No Discount              20.01   24.162098       949
1-10%                    18.92  494.334616     19790
11-20%                   16.84  502.835441     19924
21-30%                   14.87  498.463513     19917
31-40%                   12.91  507.629617     20212
41-50%                   10.95  481.015729     19208


In [40]:
result = data.groupby('Segment').agg({
    'Sales'          : 'sum',
    'Profit'         : 'sum',
    'Profit Margin %': 'mean'
}).round(2)
result[['Sales','Profit']] = result[['Sales','Profit']] / 1e6
print(result)

                 Sales      Profit  Profit Margin %
Segment                                            
Consumer   1253.297254  187.779827            14.98
Corporate  1255.143760  187.750685            14.95


In [41]:
result = data.groupby('City Type').agg({
    'Sales'          : 'sum',
    'Profit'         : 'sum',
    'Profit Margin %': 'mean'
}).sort_values('Sales', ascending=False).round(2)
result[['Sales','Profit']] = result[['Sales','Profit']] / 1e6
print(result)

                Sales      Profit  Profit Margin %
City Type                                         
Tier 1     844.420418  126.505589            14.99
Tier 2     835.435778  124.681658            14.94
Village    828.584818  124.343265            14.97


In [42]:
data.groupby('Ship Mode').agg({
    'Sales'    : 'sum',
    'Ship Days': 'mean',
    'Order ID' : 'count'
}).round(2)

,Sales,Ship Days,Order ID
Ship Mode,,,
First Class,6.283755e+08,4.00,25078
Same Day,6.272023e+08,3.99,25007
Second Class,6.271439e+08,3.97,24944
Standard Class,6.257193e+08,4.00,24971


### 📌 Discount & Segmentation Findings

- **Discounts are hurting profit significantly**
  - No discount → 20% margin
  - 41-50% discount → only 10.95% margin
  - Every 10% discount = ~2% margin loss
- **Consumer vs Corporate** — virtually identical performance
- **Tier 1 cities** lead in sales at ₹844M
- **Ship mode has no impact** on sales or delivery time
- Key recommendation: Reduce high discounts (31-50%) to protect margins
```

- Discount impact   : Clear negative effect on margins
- Segmentation      : Consumer ≈ Corporate
- City Type         : Tier 1 leads
- Ship Mode         : No significant difference

In [44]:
data.to_csv('store_data_clean.csv', index=False)
print("Clean data exported!")

Clean data exported!


In [45]:
print("=" * 45)
print("   RETAIL STORE — ANALYSIS SUMMARY")
print("=" * 45)
print(f"Total Sales    : ₹{data['Sales'].sum()/1e6:.1f}M")
print(f"Total Profit   : ₹{data['Profit'].sum()/1e6:.1f}M")
print(f"Avg Margin     : {data['Profit Margin %'].mean():.2f}%")
print(f"Total Orders   : {len(data):,}")
print(f"Top Category   : {data.groupby('Category of Goods')['Sales'].sum().idxmax()}")
print(f"Top Region     : {data.groupby('Region')['Sales'].sum().idxmax()}")
print(f"Best Month     : {data.groupby('Order Month')['Sales'].sum().idxmax()}")
print(f"Top Sub-Cat    : {data.groupby('Sub-Category')['Sales'].sum().idxmax()}")
print("=" * 45)

   RETAIL STORE — ANALYSIS SUMMARY
Total Sales    : ₹2508.4M
Total Profit   : ₹375.5M
Avg Margin     : 14.97%
Total Orders   : 100,000
Top Category   : Sessional Fruits & Vegetables
Top Region     : East
Best Month     : 5
Top Sub-Cat    : Mops
